### ISSUE TITLE
Docker for Windows

### ISSUE THREAD

**Issue by grahamehorner** [docker-library/rabbitmq#168]
Please consider creating an official docker RabbitMQ for Windows image using Windows Server Core; 

As a group of developers we use docker to create a local development environment; sharing images with others so the 'works on my machine' is a thing of the past; at present we use a number of images including including asp.net, redis, ms-sql-server, documentdb etc

**Comment by tianon** [comment]
There are installers available for Windows (https://www.rabbitmq.com/install-windows.html), so I'm not opposed, but configuration is going to be tricky.  On Linux, we use an entrypoint script which sets up the configuration and then execs `rabbitmq-server` (so the server itself is the only thing left resident).  On Windows, that pattern is not possible (due to architectural limitations).

For images like MongoDB, we've "solved" that by making the Windows images strictly bring-your-own for configuration, initialization, etc.  If an approach like that is acceptable here, then we should definitely play around with making this happen!  (and possibly whether it can work in Nano Server too)

**Comment by grahamehorner** [comment]
I think having RabbitMQ running in docker  on window server core and/or nanoserver would be a great for the inner developer loop; at present I've not been able to see how/if this could be installed given the packaged installer, this may have to be a manual install process of RabbitMQ as opposed to the provide package installers?  IMHO bring you own configuration for would be an acceptable starting point.

**Comment by grahamehorner** [comment]
FYI, found this  spring2/rabbitmq

**Comment by reecebradley** [comment]
it'd still be good if there was a rabbitmq that is under the official channel.

**Comment by michaelklishin** [comment]
@tianon IMO "bring your own configuration" is a valid approach for any OS.

There are binary releases for Windows distributed by our team as well, I'm not qualified to say which one is more suitable for the current state of provisioning and automation tooling on Windows.

**Comment by AceHack** [comment]
I created a docker nanoserver-insider image, it's very basic and need work to really be production ready. Not to mention it needs nanoserver-insider right now because erlang does not work on the current version of nanoserver.  The same dockerfile would work as well for windowsservercore.

[rabbitmq-nanoserver-insider.zip](https://github.com/docker-library/rabbitmq/files/1288934/rabbitmq-nanoserver-insider.zip)

**Comment by AceHack** [comment]
Also if the architectural limitations you speak of are not having bash then you could use windows server core and still run your bash script for windows because it can support bash now :)

https://msdn.microsoft.com/en-us/commandline/wsl/install-on-server

**Comment by KencyK** [comment]
Are there any plans of having rabbitMQ for windows? We were getting the below error:
rabbitmq_ppn_engine | {"Cookie file /var/lib/rabbitmq/.erlang.cookie must be accessible by owner only",[{auth,init_cookie,0,[{file,"auth.erl"},{line,286}]},{auth,init,1,[{file,"auth.erl"},{line,140}]},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,328}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}

**Comment by luigiberrettini** [comment]
[Here](https://github.com/EasyNetQ/EasyNetQ.Management.Client/blob/0.6.0/Tools/docker/dockerfile-rabbitmq-windows) is a working Dockerfile from which you can take inspiration.

Consider that, if you want to create a Nano Server image, you will have to deal with [Erlang issue 477](https://bugs.erlang.org/browse/ERL-477).

**Comment by luigiberrettini** [comment]
[Here](https://github.com/EasyNetQ/EasyNetQ.Management.Client/blob/v0.6.0/Tools/docker/dockerfile-rabbitmq-windows) is a working Dockerfile that can be a good start.

At the moment Nano Server images have to deal with [Erlang issue 477](https://bugs.erlang.org/browse/ERL-477).

I really prefer not to create an unofficial image: can I contribute/help in some way?

**Comment by Bhaal22** [comment]
[Rmq Windows](https://github.com/gsx-solutions/rmq-win) based on multi stage builds to startt from windowsservercore to get powershell feature and then switch to nanoservercore:1709

Need to update to 1803.

https://blog.drylm.org/posts/rabbitmq_windows_container.html

**Comment by Bhaal22** [comment]
Not so sure. 
As far as I remember I had it working on my first tries with 1709 when I did make experiments with the zip in my git repo and adding the zip via the dockerfile.

I only use multi stage builds to have powershell to download and run msiexec stuff.

**Comment by luigiberrettini** [comment]
If you look at the Dockerfile I referenced above and try to do the same with PowerShell Core as I did it won't work
This is not due to downloading or copying files, but to the Erlang installer that, even if called with the proper flags (not to use UI), executes without doing nothing

**Comment by Bhaal22** [comment]
The referenced issue is not related to run erlang installer. The attached docker file already use a multi stage build.
The referenced issue is related to erlang runtime failing.

```
Unable to load emulator DLL
(c:\erlang\erts-9.0\bin\beam.smp.dll)
Unable to load emulator DLL
(c:\erlang\erts-9.0\bin\beam.smp.dll)
```

Indeed the installer does not run as-is in the nanoserver:1709 container

**Comment by luigiberrettini** [comment]
Before installing Erlang, it could be helpful to install [ Microsoft Visual C++ Redistributable for Visual Studio 2017](https://aka.ms/vs/15/release/vc_redist.x64.exe) with `vc_redist.x64.exe /install /quiet /norestart` or `vc_redist.x64.exe /install /passive /norestart`

Unfortunately MSI is not supported on Nano Server and I think the Visual C++ Redistributable installer is an MSI installer

**Comment by grahamehorner** [comment]
@luigiberrettini MSI isn't supported for good reasons; as the packages have calls to APIs that are part of the UI layer and given the large majority of security issues related to the UI layers, the aim for nano server was be as secure and small as possible, some but not all MSIs follow the /quiet /passive rules and still try to access UI api; thank fully vc_redist isn't one of them ;) and good spot btw as I've been battling with an issue where VC libraries where missing and was getting odd messages (unrelated) but your tip made the penny drop :D

**Comment by yosifkit** [comment]
Multistage builds are not supported for Official Images: https://github.com/docker-library/official-images/issues/3383.

**Comment by luigiberrettini** [comment]
I do not understand why there is this focus on PowerShell:
 - there are bigger problems than that, like being able to install common programs, since almost none support the new WSA installer format
 - Nano Server has PowerShell Core and I was able to use it to create the RabbitMQ image if it wasn't for the Erlang installer unable to work

**Comment by Bhaal22** [comment]
Yes indeed. That's not totally correct.

the nanoservercore **container** instance has no tools other than batch-tools (cd, file, type, xcopy, ...) support.
Even the batch-tools distribution looks incomplete. I noticed the **timeout** command is missing for some reasons. And of course, all UI related libs are missing + msiexec.

So basically, with batch-tools only, you dont have such a tools like:
 * curl / Invoke-WebRequest
 * unzip / Extract-Archive
 * and maybe other tools which could be handy during the build phase of an image.

That's why maybe I mentionned too much PowerShell.

**Comment by luigiberrettini** [comment]
You are right, now I remember that I used the PowerShell Core image:
[microsoft/powershell:nanoserver](https://hub.docker.com/r/microsoft/powershell/)

This is perfectly doable without having a multistage build, but the Erlang installation problem still exist

**Comment by Bhaal22** [comment]
yes indeed. i did not had a look at the powershell core image yet.
the solution would be that erlang team releases archive of erlang-otp instead of installers ...

**Comment by luigiberrettini** [comment]
Being a "Linux program" probably the installer does not do more than unpacking and setting something in the path... even if it sets registry keys, this could be done via shell

**Comment by Bhaal22** [comment]
Indeed. I am pretty sure this is definitely overkill to have such an installer.
Will have a look at the nullsoft installer definition. Should not be too hard to understand.

**Comment by luigiberrettini** [comment]
[Here](https://github.com/erlang/otp/blob/OTP-20.3.7/erts/etc/win32/nsis/erlang.nsi) is it

It seems that it is strongly tied to having a UI experience, but luckily this mostly means writing a lot of values in the registry (install folder, shell extension and so on)

**Comment by Bhaal22** [comment]
yeah and indeed useless stuff.
I am not sure in erlang codebase they do care about registry keys.

**Comment by luigiberrettini** [comment]
Opened [ERL-635](https://bugs.erlang.org/browse/ERL-635) please vote it

**Comment by kavhad** [comment]
The link to the following Dockerfile can build an image running rabbitmq on nanoserver

https://github.com/sixeyed/dockerfiles-windows/blob/master/rabbitmq/nanoserver/sac2016-ignore/Dockerfile

It in itself depends on the following image which is using multistage build to install erlang on windows servercore but then copy resulting files into an nanoserver image

https://github.com/sixeyed/dockerfiles-windows/blob/master/erlang/nanoserver/sac2016/Dockerfile

**Comment by luigiberrettini** [comment]
I hope the Erlang team will start putting some effort also on Windows related issues instead of simply deprioritizing them and flagging as up-for-grabs or help wanted: it would be great to build and run RabbitMQ with a single stage build on microsoft/nanoserver

**Comment by xied75** [comment]
As I have already commented there in the https://bugs.erlang.org/browse/ERL-477
Erlang should be simply offered as a zip, the-Installer-mental-state is really some legacy thinking...

**Comment by luigiberrettini** [comment]
Unfortunately as you can notice from their reply they are not puttingany effort on that 😢

**Comment by tianon** [comment]
I've done a little bit of testing with this, and was able to get something preliminary running with the following `Dockerfile`.  A few issues:

- not sure how to get `RABBITMQ_DATA_DIR` to take properly
- doing `docker run IMG rabbitmq-server` doesn't work, but if I run `powershell` instead it starts up
- Ctrl+C gets captured by `cmd` (since `rabbitmq-server` is really a batch file, so it's capturing the signal and prompts for terminating the batch job instead of Erlang doing the right thing)
- the Erlang layer is almost 400MB, so there's obviously a lot of things in there that are going to be ripe for removal
- the Erlang installer is a standard NSIS installer, so it's possible we could "extract" it instead of installing it, but that's probably splitting hairs (IMO the installer is OK, we just need to clean a few things out when we're done installing to slim it down)

```dockerfile
FROM mcr.microsoft.com/windows/servercore:1803

# $ProgressPreference: https://github.com/PowerShell/PowerShell/issues/2138#issuecomment-251261324
SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]

ENV RABBITMQ_HOME C:\RabbitMQ

# PATH isn't actually set in the Docker image, so we have to set it from within the container
RUN $newPath = ('C:\erl\bin;{0}\sbin;{1}' -f $env:RABBITMQ_HOME, $env:PATH); \
	Write-Host ('Updating PATH: {0}' -f $newPath); \
	setx /M PATH $newPath
# doing this first to share cache across versions more aggressively

# no minor updates for Windows -- infeasible to rebuild from source
ENV OTP_VERSION 21.2

# pulling from erlang-solutions.com instead of erlang.org because it's a lot faster download and includes https
RUN $url = 'https://packages.erlang-solutions.com/erlang/esl-erlang/FLAVOUR_1_general/esl-erlang_{0}~windows_amd64.exe' -f $env:OTP_VERSION; \
	Write-Host ('Downloading {0} ...' -f $url); \
	Invoke-WebRequest -Uri $url -OutFile otp-setup.exe; \
	\
	Write-Host 'Installing ...'; \
	Start-Process otp-setup.exe -Wait \
		-ArgumentList @( \
# https://nsis.sourceforge.io/Which_command_line_parameters_can_be_used_to_configure_installers
			'/S', \
			'/D=C:\erl' \
		); \
	\
	Write-Host 'Removing ...'; \
	Remove-Item otp-setup.exe -Force; \
	\
	Write-Host 'Validating ...'; \
	Start-Process erl -NoNewWindow -Wait \
		-ArgumentList @( \
			'-version' \
		); \
	\
	Write-Host 'Complete.'

ENV RABBITMQ_VERSION 3.7.10

# https://www.rabbitmq.com/install-windows-manual.html
RUN $url = 'https://github.com/rabbitmq/rabbitmq-server/releases/download/v{0}/rabbitmq-server-windows-{0}.zip' -f $env:RABBITMQ_VERSION; \
	Write-Host ('Downloading {0} ...' -f $url); \
	[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; \
	Invoke-WebRequest -Uri $url -OutFile rabbit.zip; \
	\
	Write-Host 'Expanding ...'; \
	Expand-Archive -Path rabbit.zip -DestinationPath C:\; \
	\
	Write-Host 'Renaming ...'; \
	Move-Item -LiteralPath ('rabbitmq_server-{0}' -f $env:RABBITMQ_VERSION) -Destination $env:RABBITMQ_HOME; \
	\
	Write-Host 'Removing ...'; \
	Remove-Item rabbit.zip -Force; \
	\
# TODO verification
	\
	Write-Host 'Complete.'

# send all logs to TTY
ENV RABBITMQ_LOGS=- RABBITMQ_SASL_LOGS=-

# TODO RABBITMQ_DATA_DIR

# TODO CMD ["rabbitmq-server"]
CMD ["powershell"]
```

**Comment by ericjmooney** [comment]
Based on tianon's work above (which is great, btw, thank you so much) was able to get rabbit running with CMD directive like this:

CMD "cmd /k rabbitmq-server.bat"

**Comment by johnkwaters** [comment]
Wow - here I am a noob to Docker ... setting up some microservices - several dotnet 5 APIs, IdentityServer, a few mongodb... all on Windows containers. Then I spend 2 days trying to get rabbitmq to work in  docker. Little did I realize that isn't supported! Didn't even know it wasn't until I stumble across this. Should I process or just try to get it all running on Linux/WSL instead?

**Comment by ericjmooney** [comment]
@johnkwaters  If you are able, running on Linux-based containers is going to make your life a heck of a lot easier IMHO

**Comment by johnkwaters** [comment]
I hear you @ericjmooney ! I just switched - trying to iron out all the kinks, but seems way faster.

**Comment by John0King** [comment]
still no windows version ?

**Comment by tianon** [comment]
Coming back to this, I wonder how many of the limitations I noted in https://github.com/docker-library/rabbitmq/issues/168#issuecomment-459916001 are things we can actually reasonably overcome vs insurmountable. :see_no_evil:

To put that another way: I realize lots of folks _want_ a Windows-based RabbitMQ container image, but I'm not sure it really makes sense as part of what's maintained in this repository given that it will not be a comfortable or more importantly a well-supported experience. :thinking: :disappointed:

For example, not having a clean way to `docker stop` the container _gracefully_ sounds like a pretty big showstopper.

**Comment by lukebakken** [comment]
At some point I plan on migrating all of the scripts we use to manage RabbitMQ on Windows to Powershell, which will fix the `docker stop` issue.

But, I fail to see the benefit to a Windows-based RabbitMQ docker image. What would that provide to someone developing on Windows that the Linux-based one does not?

**Comment by TBBle** [comment]
The ability to run it in combination with other Windows-only containers without messing about with multiple Docker (or other runtime) instances or exploiting oddities in Docker Desktop for Windows's implementation.

Outside development, running Linux-based containers on Windows Server in production is a non-supported setup, so you'd need to fire up a Linux VM to host the RabbitMQ container on a Linux container runtime.

And all of the above workarounds are non-starters if you're trying to include a RabbitMQ instance in the same k8s Pod (or similar shared sandbox setup) as some Windows-only container, as such shared container sandboxes share the same platform/kernel.

**Comment by inikulshin** [comment]
@lukebakken 

> At some point I plan on migrating all of the scripts we use to manage RabbitMQ on Windows to Powershell, which will fix the `docker stop` issue.
> 
> But, I fail to see the benefit to a Windows-based RabbitMQ docker image. What would that provide to someone developing on Windows that the Linux-based one does not?

I'm trying to add CI action to run [EasyNetQ.Management.Client](https://github.com/EasyNetQ/EasyNetQ.Management.Client) test on .NET Framework (in addition to .NET Core). Those tests use [Docker.DotNet](https://github.com/dotnet/Docker.DotNet) package to install `rabbitmq:4.0-management` image.

However,  .NET Framework tests can't run on `ubuntu-latest`, and all [available Windows runners](https://github.com/actions/runner-images?tab=readme-ov-file#available-images) are Windows Server, that don't support Docker Desktop with Linux containers.

**Comment by lukebakken** [comment]
@inikulshin - please feel free to use the same scripts I do to install Erlang and RabbitMQ on Windows - https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/.ci/windows/gha-setup.ps1

**Comment by lukebakken** [comment]
I haven't found the time to work on RabbitMQ's init scripts for Windows and as far as I know no Windows users are willing to step up to fund this work, so I'm going to close this issue as not planned. It has been over 8 years since the initial request.
